sematic/external_resource.py

```diff
@@ -0,0 +1,224 @@
+# Standard Library
+import time
+import uuid
+from dataclasses import dataclass, field
+from enum import Enum, unique
+from typing import final
+
+# Sematic
+from sematic.future_context import AbstractExternalResource
+from sematic.utils.exceptions import IllegalStateTransitionError
+
+
+@unique
+class ResourceState(Enum):
+    """Represents the state of an external resource
+
+    Attributes
+    ----------
+    CREATED:
+        The object representing the resource has been instantiated in-memory
+        but no logic to allocate the resource has been executed.
+    ACTIVATING:
```

Comment:
We almost certainly want these transitional states for some things. Being able to tell when we asked for the thing to be allocated and when it was actually ready for use will be important. Likewise in the other direction. For the hello-world-y sample I have this is overkill, but realistic external compute always has some non-trivial latency to spin up/down.